src/e2e/java/teammates/e2e/util/JMeterConfig.java

```diff
@@ -122,26 +135,64 @@ public HashTree createTestPlan() {
+        if (isPostEndpoint) {
+            regexExtractor = new RegexExtractor();
+            regexExtractor.setName("Regular Expression Extractor");
+            regexExtractor.setUseField("true"); // Find matches in response headers
+            regexExtractor.setRefName("csrfToken");
+            regexExtractor.setRegex("Set-Cookie: CSRF-TOKEN=(.+?);");
+            regexExtractor.setTemplate("$1$");
+            regexExtractor.setProperty(TestElement.TEST_CLASS, RegexExtractor.class.getName());
+            regexExtractor.setProperty(TestElement.GUI_CLASS, RegexExtractorGui.class.getName());
+        }
+
         // Test API Endpoint HTTP Request
         HTTPSamplerProxy apiSampler = new HTTPSamplerProxy();
         apiSampler.setName("Test Endpoint");
         apiSampler.setPath(getTestEndpoint());
         apiSampler.setMethod(getTestMethod());
-        getTestArguments().forEach((key, value) -> {
+        getTestEndpointParameters().forEach((key, value) -> {
             apiSampler.addArgument(key, value);
         });
         apiSampler.setEnabled(true);
         apiSampler.setProperty(TestElement.TEST_CLASS, HTTPSamplerProxy.class.getName());
         apiSampler.setProperty(TestElement.GUI_CLASS, HttpTestSampleGui.class.getName());
 
+        // Add HTTP POST Body Data
+        if (isPostEndpoint && !getTestEndpointPostVariables().isEmpty()) {
+            getTestEndpointPostVariables().forEach(val -> {
+                apiSampler.addNonEncodedArgument("", val, "");
+                apiSampler.setPostBodyRaw(true);
+            });
+        }
+
+        // HTTP Header Manager
+        HeaderManager headerManager = null;
+        if (isPostEndpoint) {
+            headerManager = new HeaderManager();
+            headerManager.setName("HTTP Header Manager");
+            headerManager.add(new Header("Content-Type", "text/plain"));
+            headerManager.add(new Header("X-CSRF-TOKEN", "${csrfToken}"));
+            headerManager.setProperty(TestElement.TEST_CLASS, HeaderManager.class.getName());
+            headerManager.setProperty(TestElement.GUI_CLASS, HeaderPanel.class.getName());
+        }
+
         // Create Test plan
         HashTree testPlanHashTree = new ListedHashTree();
         HashTree threadGroupHashTree = testPlanHashTree.add(testPlan, threadGroup);
+        // Add Config elements
         threadGroupHashTree.add(csvDataSet);
         threadGroupHashTree.add(cookieManager);
         threadGroupHashTree.add(defaultSampler);
-        threadGroupHashTree.add(onceOnlyController, loginSampler);
-        threadGroupHashTree.add(apiSampler);
+        // Add HTTP samplers (and configurations, pre/post processors)
+        HashTree loginSamplerHashTree = threadGroupHashTree.add(onceOnlyController, loginSampler);
+        HashTree apiSamplerHashTree = threadGroupHashTree.add(apiSampler);
+        if (isPostEndpoint) {
+            loginSamplerHashTree.add(regexExtractor);
+            apiSamplerHashTree.add(headerManager);
```

Comment:
Rather than initializing `regexExtractor` and `headerManager` far before this if block, why not initialize everything here? That way you also don't need to repeat the `is (postEndpoint)` condition check.